In [3]:
# Bibek Shiwakoti
# 03/25/2023

In [4]:
# Let's practice associated rules in python.

In [5]:
# Apriori Principal implementation

In [6]:
# pip install mlxtend

In [7]:
dataset = [
    ['Milk', 'Onion', 'Nutmeg', 'Kidney Beans', 'Eggs', 'Yoghurt'],
    ['Dill', 'Onion', 'Nutmeg', 'Kidney Beans', 'Eggs', 'Yoghurt'],
    ['Milk', 'Apple','Kidney Beans', 'Eggs'] ,
    ['Milk', 'Unicorn', 'Corn', 'Kidney Beans','Yoghurt']  ,
    ['Corn', 'Onion', 'Onion', 'Kidney Beans','Ice Cream', 'Eggs']
]

In [28]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder

In [9]:
dataset

[['Milk', 'Onion', 'Nutmeg', 'Kidney Beans', 'Eggs', 'Yoghurt'],
 ['Dill', 'Onion', 'Nutmeg', 'Kidney Beans', 'Eggs', 'Yoghurt'],
 ['Milk', 'Apple', 'Kidney Beans', 'Eggs'],
 ['Milk', 'Unicorn', 'Corn', 'Kidney Beans', 'Yoghurt'],
 ['Corn', 'Onion', 'Onion', 'Kidney Beans', 'Ice Cream', 'Eggs']]

In [10]:
te =TransactionEncoder()
te_aray = te.fit(dataset).transform(dataset)

In [11]:
df =pd.DataFrame(te_aray,columns=te.columns_)
df

Apple   Corn   Dill   Eggs  Ice Cream  Kidney Beans   Milk  Nutmeg  Onion  \
0  False  False  False   True      False          True   True    True   True   
1  False  False   True   True      False          True  False    True   True   
2   True  False  False   True      False          True   True   False  False   
3  False   True  False  False      False          True   True   False  False   
4  False   True  False   True       True          True  False   False   True   

   Unicorn  Yoghurt  
0    False     True  
1    False     True  
2    False    False  
3     True     True  
4    False    False

There are now 11 items in the list and the data encoded is shown in above dataframe

In [12]:
from mlxtend.frequent_patterns import apriori

Printing the frequent itemsets with their support

In [13]:
frequent_itemsets = apriori(df,min_support=0.6,use_colnames=False)
frequent_itemsets

support   itemsets
0       0.8        (3)
1       1.0        (5)
2       0.6        (6)
3       0.6        (8)
4       0.6       (10)
5       0.8     (3, 5)
6       0.6     (8, 3)
7       0.6     (5, 6)
8       0.6     (8, 5)
9       0.6    (10, 5)
10      0.6  (8, 3, 5)

## Association Rule Mining

In [21]:
from mlxtend.frequent_patterns import association_rules
res = association_rules(frequent_itemsets,metric='confidence', min_threshold=0.8)


In [22]:
res

antecedents consequents  antecedent support  consequent support  support  \
0         (3)         (5)                 0.8                 1.0      0.8   
1         (5)         (3)                 1.0                 0.8      0.8   
2         (8)         (3)                 0.6                 0.8      0.6   
3         (6)         (5)                 0.6                 1.0      0.6   
4         (8)         (5)                 0.6                 1.0      0.6   
5        (10)         (5)                 0.6                 1.0      0.6   
6      (8, 3)         (5)                 0.6                 1.0      0.6   
7      (8, 5)         (3)                 0.6                 0.8      0.6   
8         (8)      (3, 5)                 0.6                 0.8      0.6   

   confidence  lift  leverage  conviction  zhangs_metric  
0         1.0  1.00      0.00         inf            0.0  
1         0.8  1.00      0.00         1.0            0.0  
2         1.0  1.25      0.12         inf            0.5  
3         1.0  1.00      0.00         inf            0.0  
4         1.0  1.00      0.00         inf            0.0  
5         1.0  1.00      0.00         inf            0.0  
6         1.0  1.00      0.00         inf            0.0  
7         1.0  1.25      0.12         inf            0.5  
8         1.0  1.25      0.12         inf            0.5

In [23]:
res_1 = res[['antecedents','consequents','support','confidence','lift']]

In [24]:
res_1

antecedents consequents  support  confidence  lift
0         (3)         (5)      0.8         1.0  1.00
1         (5)         (3)      0.8         0.8  1.00
2         (8)         (3)      0.6         1.0  1.25
3         (6)         (5)      0.6         1.0  1.00
4         (8)         (5)      0.6         1.0  1.00
5        (10)         (5)      0.6         1.0  1.00
6      (8, 3)         (5)      0.6         1.0  1.00
7      (8, 5)         (3)      0.6         1.0  1.25
8         (8)      (3, 5)      0.6         1.0  1.25

In [25]:
res_2 = res_1[res_1['confidence']>=1]
res_2

antecedents consequents  support  confidence  lift
0         (3)         (5)      0.8         1.0  1.00
2         (8)         (3)      0.6         1.0  1.25
3         (6)         (5)      0.6         1.0  1.00
4         (8)         (5)      0.6         1.0  1.00
5        (10)         (5)      0.6         1.0  1.00
6      (8, 3)         (5)      0.6         1.0  1.00
7      (8, 5)         (3)      0.6         1.0  1.25
8         (8)      (3, 5)      0.6         1.0  1.25

By doing this implementation we have learned how apriori algorithm works clearly.

*********************************************************************

## Now we are going to implement a FP based algorithm

In [26]:
# looking at the execution time of previous algorithm i.e. Apriori

In [30]:
from mlxtend.frequent_patterns import apriori
%timeit apriori (df,min_support=0.6)

2.63 ms ± 205 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [32]:
from mlxtend.frequent_patterns import fpgrowth
%timeit fpgrowth(df,min_support=0.6)

955 µs ± 31.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


## Conclusion

for the same dataset we have applied the apriori and fpgrowth algorithm and after calculating the computing time apriori algorithm took 2.63 ms whereas fpgrowth algorithm takes only 955 microsecond (µs ). Hence we found that fpgrowth algorithm is more efficient in terms of computation time it is way faster than apriori algorithm.

## Thank You